### PR TITLE
BlockDropItemEvent

### PR DESCRIPTION
**Pull request by TheAbsolutionism**
### Description
This PR aims to add the `BlockDropItemEvent`

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** #5959

### CODE REVIEW

File: src/main/java/ch/njol/skript/events/EvtBlock.java [hunk 1]
```diff
@@ -74,6 +70,22 @@ public class EvtBlock extends SkriptEvent {
...
+		Skript.registerEvent("Block Drop", EvtBlock.class, BlockDropItemEvent.class, "block drop[ping] [[of] %-itemtypes/blockdatas%]")
+			.description(
+				"Called when a block broken by a player drops items",
+				"<ul>",
+				"<li>event-player : The player that broke the block</li>",
+				"<li>past event-block : The block that was broken</li>",
+				"<li>event-block : The block after being broken</li>",
+				"<li>event-items (or drops) : The drops of the block</li>",
+				"<li>event-entities : The entities of the dropped items</li>",
+				"</ul>",
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/events/EvtBlock.java:82]: is there a precedent for this?

File: src/main/java/ch/njol/skript/events/EvtBlock.java [hunk 2]
```diff
@@ -74,6 +70,22 @@ public class EvtBlock extends SkriptEvent {
 			.description("Called when a block is created, but not by a player, e.g. snow forms due to snowfall, water freezes in cold biomes. This isn't called when block spreads (mushroom growth, water physics etc.), as it has its own event (see <a href='#spread'>spread event</a>).")
 			.examples("on form of snow:", "on form of a mushroom:")
 			.since("1.0, 2.6 (BlockData support)");
+		Skript.registerEvent("Block Drop", EvtBlock.class, BlockDropItemEvent.class, "block drop[ping] [[of] %-itemtypes/blockdatas%]")
+			.description(
+				"Called when a block broken by a player drops items",
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/events/EvtBlock.java:75]:
```suggestion
				"Called when a block broken by a player drops something.",
```

File: src/main/java/ch/njol/skript/events/EvtBlock.java [hunk 3]
```diff
@@ -74,6 +70,22 @@ public class EvtBlock extends SkriptEvent {
...
+				"<ul>",
+				"<li>event-player : The player that broke the block</li>",
+				"<li>past event-block : The block that was broken</li>",
+				"<li>event-block : The block after being broken</li>",
+				"<li>event-items (or drops) : The drops of the block</li>",
+				"<li>event-entities : The entities of the dropped items</li>",
+				"</ul>",
+				"",
+				"If the breaking of the block leads to others being broken, such as torches, they will appear",
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/events/EvtBlock.java:84]: appear where?

File: src/main/java/ch/njol/skript/events/EvtBlock.java [hunk 4]
```diff
@@ -74,6 +70,22 @@ public class EvtBlock extends SkriptEvent {
...
+				"in \"event-items\" and \"event-entities\"."
+			)
+			.examples("on block drop:", "on block drop of oak log:")
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/events/EvtBlock.java:87]: should have at least some code in the event itself like broadcast, doesnt matter

File: src/main/java/ch/njol/skript/expressions/ExprDrops.java [hunk 5]
```diff
@@ -61,28 +64,32 @@ public class ExprDrops extends SimpleExpression<ItemType> {
 
 	@Override
 	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
-		if (!getParser().isCurrentEvent(EntityDeathEvent.class)) {
-			Skript.error("The expression 'drops' can only be used in death events", ErrorQuality.SEMANTIC_ERROR);
+		if (!getParser().isCurrentEvent(EntityDeathEvent.class, BlockDropItemEvent.class)) {
+			Skript.error("The expression 'drops' can only be used in death events and block drop events", ErrorQuality.SEMANTIC_ERROR);
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprDrops.java:68]:
```suggestion
			Skript.error("The expression 'drops' can only be used in death events and block drop events");
``` 
semantic is default

**Review by Efnilite — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprDrops.java [hunk 6]
```diff
@@ -144,20 +161,20 @@ public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
 		// todo: All the experience stuff should be removed from this class for 2.8 and given to ExprExperience
 
 		// handle experience
-		if (deltaExperience > -1) {
+		if (deltaExperience > -1 && event instanceof EntityDeathEvent entityDeathEvent) {
 			switch (mode) {
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprDrops.java:165]: enhanced switch

**Review by Efnilite — APPROVED** [review]

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 7]
```diff
@@ -1935,5 +1936,31 @@ public RegainReason get(EntityRegainHealthEvent event) {
 				return event.getRegainReason();
 			}
 		}, EventValues.TIME_NOW);
+
+		// BlockDropItemEvent
+		EventValues.registerEventValue(BlockDropItemEvent.class, Block.class, new Getter<Block, BlockDropItemEvent>() {
+			@Override
+			public @Nullable Block get(BlockDropItemEvent event) {
+				return new BlockStateBlock(event.getBlockState());
+			}
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:1945]: remove nullable if can't return null

File: src/main/java/ch/njol/skript/expressions/ExprDrops.java [hunk 8]
```diff
@@ -61,28 +64,32 @@ public class ExprDrops extends SimpleExpression<ItemType> {
...
-	@Nullable
-	protected ItemType[] get(Event e) {
-		if (!(e instanceof EntityDeathEvent))
-			return null;
-
-		return ((EntityDeathEvent) e).getDrops()
-			.stream()
-			.map(ItemType::new)
-			.toArray(ItemType[]::new);
+	protected @Nullable ItemType[] get(Event event) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprDrops.java:75]:
```suggestion
	protected ItemType @Nullable [] get(Event event) {
```

File: src/main/java/ch/njol/skript/expressions/ExprDrops.java [hunk 9]
```diff
@@ -144,24 +161,17 @@ public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
...
-				case ADD:
-					((EntityDeathEvent) event).setDroppedExp(originalExperience + deltaExperience);
-					break;
-				case REMOVE:
-					((EntityDeathEvent) event).setDroppedExp(originalExperience - deltaExperience);
-					// fallthrough to check for removeAllExperience
-				case REMOVE_ALL:
+				case SET -> {entityDeathEvent.setDroppedExp(deltaExperience);}
+				case ADD -> {entityDeathEvent.setDroppedExp(originalExperience + deltaExperience);}
+				case REMOVE -> {entityDeathEvent.setDroppedExp(originalExperience - deltaExperience);} // fallthrough to check for removeAllExperience
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprDrops.java:168]: this doesn't fall through anymore

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/events/EvtBlock.java [hunk 10]
```diff
@@ -144,8 +164,8 @@ else if (o instanceof BlockData && finalBlockData != null)
 	}
 	
 	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
-		return "break/place/burn/fade/form of " + Classes.toString(types);
+	public @NotNull String toString(@Nullable Event event, boolean debug) {
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/events/EvtBlock.java:167]: This `@NotNull` isn't needed at this position when Skript is the one to call this, and Skript already asserts it not null.
Same with import for this annotation.

File: src/main/java/ch/njol/skript/expressions/ExprDrops.java [hunk 11]
```diff
@@ -101,12 +108,22 @@ public Class<?>[] acceptChange(ChangeMode mode) {
 	}
 
 	@Override
-	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
-		if (!(event instanceof EntityDeathEvent))
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprDrops.java:112]:
```suggestion
```

File: src/main/java/ch/njol/skript/expressions/ExprDrops.java [hunk 12]
```diff
@@ -61,28 +64,32 @@ public class ExprDrops extends SimpleExpression<ItemType> {
...
+				.stream()
+				.map(ItemType::new)
+				.toArray(ItemType[]::new);
+		} else if (event instanceof BlockDropItemEvent blockDropItemEvent) {
+			return blockDropItemEvent.getItems()
+				.stream()
+				.map(Item::getItemStack)
+				.map(ItemType::new)
+				.toArray(ItemType[]::new);
+		}
```
Comment (TheLimeGlass) [inline src/main/java/ch/njol/skript/expressions/ExprDrops.java:87]:
```suggestion
		}
		assert false;
```
Optional as there aren't any tests to catch this.

**Review by TheLimeGlass — APPROVED** [review]

File: src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java [hunk 13]
```diff
@@ -1935,5 +1936,31 @@ public RegainReason get(EntityRegainHealthEvent event) {
...
+			@Override
+			public @Nullable ItemStack[] get(BlockDropItemEvent event) {
+				return event.getItems().stream().map(Item::getItemStack).toArray(ItemStack[]::new);
+			}
+		}, EventValues.TIME_NOW);
+		EventValues.registerEventValue(BlockDropItemEvent.class, Entity[].class, new Getter<Entity[], BlockDropItemEvent>() {
+			@Override
+			public @Nullable Entity[] get(BlockDropItemEvent event) {
+				return event.getItems().toArray(Entity[]::new);
+			}
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/classes/data/BukkitEventValues.java:1963]: still need nullables addressed

File: src/main/java/ch/njol/skript/expressions/ExprDrops.java (hunk 9 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/expressions/ExprDrops.java:167]: unresolved?

**Review by sovdeeth — CHANGES_REQUESTED** [review]

**Review by sovdeeth — APPROVED** [review]

**Pull request merged**